File: src/lib/api/reports.ts
import { apiDelete, apiGet, apiPost, apiPut } from "$lib/api/client";

export type GroupBy = "day" | "month" | "quarter" | "year";

export type CashflowReportInput = {
  book_id: number;
  date_from: string | null;
  date_to: string | null;
  group_by: GroupBy | null;
};

export type CashflowRow = {
  period_start: string;
  inflow_minor: number;
  outflow_minor: number;
  net_minor: number;
};

export type NetWorthReportInput = {
  book_id: number;
  date_from: string | null;
  date_to: string | null;
  group_by: GroupBy | null;
};

export type NetWorthRow = {
  period_start: string;
  assets_minor: number;
  liabilities_minor: number;
  net_worth_minor: number;
};

export type AccountTrendReportInput = {
  book_id: number;
  account_ids: number[] | null;
  date_from: string | null;
  date_to: string | null;
  group_by: GroupBy | null;
};

export type AccountTrendRow = {
  period_start: string;
  account_id: number;
  account_name: string;
  balance_minor: number;
};

export type CategorySpendReportInput = {
  book_id: number;
  date_from: string | null;
  date_to: string | null;
  category_ids: number[] | null;
};

export type CategorySpendRow = {
  category_id: number;
  category_name: string;
  total_minor: number;
};

export type PayeeTotalsReportInput = {
  book_id: number;
  date_from: string | null;
  date_to: string | null;
  payee_ids: number[] | null;
};

export type PayeeTotalRow = {
  payee_id: number;
  payee_name: string;
  total_minor: number;
};

export type RealizedGainEntry = {
  tx_id: number;
  txn_date: string;
  commodity_id: number;
  quantity_minor: number;
  proceeds_minor: number;
  quote_commodity_id: number | null;
  cost_basis_minor: number;
  gain_loss_minor: number;
  proceeds_missing: boolean;
};

export type UnrealizedGainEntry = {
  account_id: number;
  account_name: string;
  account_type: string;
  commodity_id: number;
  commodity_name: string;
  value_minor: number;
  cost_basis_minor: number;
  unrealized_gain_minor: number;
  price_missing: boolean;
};

export type PortfolioPerformanceSummary = {
  book_id: number;
  base_commodity_id: number;
  as_of_date: string | null;
  market_value_minor: number;
  cost_basis_minor: number;
  unrealized_gain_minor: number;
  realized_gain_minor: number;
  income_minor: number;
  price_missing_count: number;
};

export type AccountValuationRow = {
  account_id: number;
  account_name: string;
  account_commodity_id: number;
  market_value_minor: number;
  cost_basis_minor: number;
  unrealized_gain_minor: number;
  price_missing: boolean;
  fx_missing: boolean;
};

export type CurrencyExposureRow = {
  commodity_id: number;
  commodity_name: string;
  quantity_minor: number;
  native_value_minor: number;
  price_missing: boolean;
};

export type ReportDefinitionCreateInput = {
  book_id: number;
  name: string;
  kind: string;
  query_type: string;
  query_text: string;
  params_schema: string | null;
};

export type ReportDefinitionUpdateInput = ReportDefinitionCreateInput & {
  id: number;
};

export type ReportDefinitionSummary = {
  id: number;
  book_id: number;
  name: string;
  kind: string;
  query_type: string;
  query_text: string;
  params_schema: string | null;
  created_at: string;
};

export type ReportRunCreateInput = {
  book_id: number;
  definition_id: number;
  params_hash: string;
  as_of_seq: number;
  pricing_mode: "frozen" | "latest_corrected";
  pricing_policy_id: number | null;
  pricing_policy_version: string | null;
  valuation_snapshot_id: number | null;
  pricing_resolved_at: string | null;
  result_json: string;
};

export type ReportRunSummary = {
  id: number;
  book_id: number;
  definition_id: number;
  params_hash: string;
  as_of_seq: number;
  pricing_mode: "frozen" | "latest_corrected";
  pricing_policy_id: number | null;
  pricing_policy_version: string | null;
  valuation_snapshot_id: number | null;
  pricing_resolved_at: string | null;
  result_json: string;
  created_at: string;
};

function buildQuery(params: Record<string, string | number | null | undefined>): string {
  const searchParams = new URLSearchParams();

  for (const [key, value] of Object.entries(params)) {
    if (value !== null && value !== undefined) {
      searchParams.set(key, String(value));
    }
  }

  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export async function reportCashflow(input: CashflowReportInput): Promise<CashflowRow[]> {
  return apiPost<CashflowRow[], CashflowReportInput>("/reports/cashflow", input);
}

export async function reportNetWorth(input: NetWorthReportInput): Promise<NetWorthRow[]> {
  return apiPost<NetWorthRow[], NetWorthReportInput>("/reports/net-worth", input);
}

export async function reportAccountTrends(input: AccountTrendReportInput): Promise<AccountTrendRow[]> {
  return apiPost<AccountTrendRow[], AccountTrendReportInput>("/reports/account-trends", input);
}

export async function reportCategorySpend(input: CategorySpendReportInput): Promise<CategorySpendRow[]> {
  return apiPost<CategorySpendRow[], CategorySpendReportInput>("/reports/category-spend", input);
}

export async function reportPayeeTotals(input: PayeeTotalsReportInput): Promise<PayeeTotalRow[]> {
  return apiPost<PayeeTotalRow[], PayeeTotalsReportInput>("/reports/payee-totals", input);
}

export async function realizedGainsReport(
  bookId: number,
  dateFrom: string | null,
  dateTo: string | null,
  costBasisProfileId?: number | null
): Promise<RealizedGainEntry[]> {
  const path = `/reports/realized-gains${buildQuery({
    book_id: bookId,
    date_from: dateFrom,
    date_to: dateTo,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<RealizedGainEntry[]>(path);
}

export async function unrealizedGainsReport(
  bookId: number,
  baseCommodityId: number,
  asOfDate: string | null,
  costBasisProfileId?: number | null
): Promise<UnrealizedGainEntry[]> {
  const path = `/reports/unrealized-gains${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<UnrealizedGainEntry[]>(path);
}

export async function investmentPerformanceReport(
  bookId: number,
  baseCommodityId: number,
  asOfDate: string | null,
  costBasisProfileId?: number | null
): Promise<PortfolioPerformanceSummary> {
  const path = `/reports/investment-performance${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<PortfolioPerformanceSummary>(path);
}

export async function accountValuationReport(
  bookId: number,
  baseCommodityId: number,
  asOfDate: string | null,
  costBasisProfileId?: number | null
): Promise<AccountValuationRow[]> {
  const path = `/reports/account-valuation${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<AccountValuationRow[]>(path);
}

export async function currencyExposureReport(bookId: number, asOfDate: string | null): Promise<CurrencyExposureRow[]> {
  const path = `/reports/currency-exposure${buildQuery({ book_id: bookId, as_of_date: asOfDate })}`;
  return apiGet<CurrencyExposureRow[]>(path);
}

export async function listReportDefinitions(bookId: number): Promise<ReportDefinitionSummary[]> {
  return apiGet<ReportDefinitionSummary[]>(`/reports/definitions${buildQuery({ book_id: bookId })}`);
}

export async function getReportDefinition(definitionId: number): Promise<ReportDefinitionSummary> {
  return apiGet<ReportDefinitionSummary>(`/reports/definitions/${definitionId}`);
}

export async function saveReportDefinition(
  input: ReportDefinitionCreateInput | ReportDefinitionUpdateInput
): Promise<ReportDefinitionSummary> {
  if ("id" in input) {
    return apiPut<ReportDefinitionSummary, ReportDefinitionUpdateInput>(`/reports/definitions/${input.id}`, input);
  }
  return apiPost<ReportDefinitionSummary, ReportDefinitionCreateInput>("/reports/definitions", input);
}

export async function deleteReportDefinition(definitionId: number): Promise<void> {
  return apiDelete<void>(`/reports/definitions/${definitionId}`);
}

export async function listReportRuns(bookId: number, definitionId?: number | null): Promise<ReportRunSummary[]> {
  return apiGet<ReportRunSummary[]>(
    `/reports/runs${buildQuery({ book_id: bookId, definition_id: definitionId ?? null })}`
  );
}

export async function createReportRun(input: ReportRunCreateInput): Promise<ReportRunSummary> {
  return apiPost<ReportRunSummary, ReportRunCreateInput>("/reports/runs", input);
}

File: apps/api/src/rekenraam_api/api/v1/reports.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status

from rekenraam_api.api.dependencies import get_investment_service, get_report_service
from rekenraam_api.schemas.investments import (
    AccountValuationRow,
    CorporateActionSummary,
    CurrencyExposureRow,
    PortfolioPerformanceQuery,
    PortfolioPerformanceSummary,
    PositionsQuery,
    RealizedGainEntry,
    RealizedGainsQuery,
    UnrealizedGainEntry,
    UnrealizedGainsQuery,
)
from rekenraam_api.schemas.reports import (
    CashflowReportInput,
    CashflowRow,
    AccountTrendReportInput,
    AccountTrendRow,
    CategorySpendReportInput,
    CategorySpendRow,
    NetWorthReportInput,
    NetWorthRow,
    PayeeTotalsReportInput,
    PayeeTotalRow,
    ReportDefinitionCreateInput,
    ReportDefinitionSummary,
    ReportDefinitionUpdateInput,
    ReportRunCreateInput,
    ReportRunSummary,
)
from rekenraam_api.services.investments import InvestmentService
from rekenraam_api.services.reports import ReportService


router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/definitions", response_model=list[ReportDefinitionSummary])
async def list_report_definitions(
    book_id: int = Query(default=1),
    report_service: ReportService = Depends(get_report_service),
) -> list[ReportDefinitionSummary]:
    return await report_service.list_report_definitions(book_id)


@router.get("/definitions/{definition_id}", response_model=ReportDefinitionSummary)
async def get_report_definition(
    definition_id: int,
    report_service: ReportService = Depends(get_report_service),
) -> ReportDefinitionSummary:
    item = await report_service.get_report_definition(definition_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="report definition not found")
    return item


@router.post("/definitions", response_model=ReportDefinitionSummary)
async def create_report_definition(
    input: ReportDefinitionCreateInput,
    report_service: ReportService = Depends(get_report_service),
) -> ReportDefinitionSummary:
    try:
        return await report_service.create_report_definition(input)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error


@router.put("/definitions/{definition_id}", response_model=ReportDefinitionSummary)
async def update_report_definition(
    definition_id: int,
    input: ReportDefinitionUpdateInput,
    report_service: ReportService = Depends(get_report_service),
) -> ReportDefinitionSummary:
    if definition_id != input.id:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="path id must match body id")
    try:
        item = await report_service.update_report_definition(input)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="report definition not found")
    return item


@router.delete("/definitions/{definition_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_report_definition(
    definition_id: int,
    report_service: ReportService = Depends(get_report_service),
) -> None:
    deleted = await report_service.delete_report_definition(definition_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="report definition not found")


@router.get("/runs", response_model=list[ReportRunSummary])
async def list_report_runs(
    book_id: int = Query(default=1),
    definition_id: int | None = Query(default=None),
    report_service: ReportService = Depends(get_report_service),
) -> list[ReportRunSummary]:
    return await report_service.list_report_runs(book_id, definition_id)


@router.post("/runs", response_model=ReportRunSummary)
async def create_report_run(
    input: ReportRunCreateInput,
    report_service: ReportService = Depends(get_report_service),
) -> ReportRunSummary:
    try:
        return await report_service.create_report_run(input)
    except ValueError as error:
        detail = str(error)
        if detail == "report definition not found":
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=detail) from error
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=detail) from error


@router.post("/cashflow", response_model=list[CashflowRow])
async def report_cashflow(
    input: CashflowReportInput,
    report_service: ReportService = Depends(get_report_service),
) -> list[CashflowRow]:
    try:
        return await report_service.report_cashflow(input)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error


@router.post("/net-worth", response_model=list[NetWorthRow])
async def report_net_worth(
    input: NetWorthReportInput,
    report_service: ReportService = Depends(get_report_service),
) -> list[NetWorthRow]:
    try:
        return await report_service.report_net_worth(input)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error


@router.post("/account-trends", response_model=list[AccountTrendRow])
async def report_account_trends(
    input: AccountTrendReportInput,
    report_service: ReportService = Depends(get_report_service),
) -> list[AccountTrendRow]:
    try:
        return await report_service.report_account_trends(input)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error


@router.post("/category-spend", response_model=list[CategorySpendRow])
async def report_category_spend(
    input: CategorySpendReportInput,
    report_service: ReportService = Depends(get_report_service),
) -> list[CategorySpendRow]:
    return await report_service.report_category_spend(input)


@router.post("/payee-totals", response_model=list[PayeeTotalRow])
async def report_payee_totals(
    input: PayeeTotalsReportInput,
    report_service: ReportService = Depends(get_report_service),
) -> list[PayeeTotalRow]:
    return await report_service.report_payee_totals(input)


@router.get("/realized-gains", response_model=list[RealizedGainEntry])
async def report_realized_gains(
    book_id: int = Query(default=1),
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    cost_basis_profile_id: int | None = Query(default=None),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> list[RealizedGainEntry]:
    return await investment_service.report_realized_gains(
        RealizedGainsQuery(
            book_id=book_id,
            date_from=date_from,
            date_to=date_to,
            cost_basis_profile_id=cost_basis_profile_id,
        )
    )


@router.get("/unrealized-gains", response_model=list[UnrealizedGainEntry])
async def report_unrealized_gains(
    book_id: int = Query(default=1),
    base_commodity_id: int = Query(...),
    as_of_date: date | None = Query(default=None),
    cost_basis_profile_id: int | None = Query(default=None),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> list[UnrealizedGainEntry]:
    return await investment_service.report_unrealized_gains(
        UnrealizedGainsQuery(
            book_id=book_id,
            base_commodity_id=base_commodity_id,
            as_of_date=as_of_date,
            cost_basis_profile_id=cost_basis_profile_id,
        )
    )


@router.get("/investment-performance", response_model=PortfolioPerformanceSummary)
async def report_investment_performance(
    book_id: int = Query(default=1),
    base_commodity_id: int = Query(...),
    as_of_date: date | None = Query(default=None),
    cost_basis_profile_id: int | None = Query(default=None),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> PortfolioPerformanceSummary:
    return await investment_service.portfolio_performance(
        PortfolioPerformanceQuery(
            book_id=book_id,
            base_commodity_id=base_commodity_id,
            as_of_date=as_of_date,
            cost_basis_profile_id=cost_basis_profile_id,
        )
    )


@router.get("/account-valuation", response_model=list[AccountValuationRow])
async def report_account_valuation(
    book_id: int = Query(default=1),
    base_commodity_id: int = Query(...),
    as_of_date: date | None = Query(default=None),
    cost_basis_profile_id: int | None = Query(default=None),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> list[AccountValuationRow]:
    return await investment_service.account_valuation(
        PortfolioPerformanceQuery(
            book_id=book_id,
            base_commodity_id=base_commodity_id,
            as_of_date=as_of_date,
            cost_basis_profile_id=cost_basis_profile_id,
        )
    )


@router.get("/currency-exposure", response_model=list[CurrencyExposureRow])
async def report_currency_exposure(
    book_id: int = Query(default=1),
    as_of_date: date | None = Query(default=None),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> list[CurrencyExposureRow]:
    return await investment_service.currency_exposure(PositionsQuery(book_id=book_id, as_of_date=as_of_date))


@router.get("/corporate-actions", response_model=list[CorporateActionSummary])
async def report_corporate_actions(
    book_id: int = Query(default=1),
    investment_service: InvestmentService = Depends(get_investment_service),
) -> list[CorporateActionSummary]:
    return await investment_service.list_corporate_actions(book_id)
